### ISSUE TITLE
Summary virtiofsd known issues

### ISSUE THREAD

**Issue by jcvenegas** [kata-containers/runtime#2797]
# Description of problem

This issue is a high-level summary of known issues.

Today two implementations use virtiofsd as shared file system:
- qemu-virtiofs 
  master: only docker is tested http://jenkins.katacontainers.io/job/kata-containers-tests-ubuntu-18-04-virtiofs-PR/
  metrics: http://jenkins.katacontainers.io/job/kata-metrics-runtime-ubuntu-virtiofs-18-04-PR/

Coverage testing (List of tests that run today with it)
  - TODO checklist of test that runs today
- cloud-hypervisor (all jobs run with virtiofs
Coverage testing (List of tests that run today with it)
  - docker
  - k8s (crio, contained)  kata tests and end-to-end
  - podman
  - Metrics CI (blog bench)
  - todo:
     posix file system testing

Based on the coverage of test listed above the following bugs has been found:
Filesystem:
- [ ]  qemu+virtiofsd: Blog bench test hangs
    **Status**: virtio_fs_cache = "none" fix the issue
    kata-containers/tests#2717 
- [ ]  Package manager issues because missing extended attributes
    **Status**: Today not default but can be added in config.toml
    kata-containers/tests#2008 package manager update test failing on virtiofs CI

- [ ]  #2598 virtiofs to NFS volume shows strange "access denied" errors bug needs-review

- [ ]  #2770 deleting a mapped file from host doesn't allow it to be recreated
    **status**: Should be fixed, need confirmation similiar issue found in k8s e2e.

- [ ]  #1888 Unable to start docker in docker with virtio-fs bug needs-help
     **status**: failed to work with ovelayfs, fixed, need a test to confirm and keep testing
- [ ] #1505 9P/virtio-fs: inotify does not work: ConfigMap Updates limitation
      **status**: limitation
- [ ] Run posix file system testing

Devices:
- [ ]  #1923 vm stop when virtiofs + NIC passthrough bug needs-review
    **status**: Need test
- [ ]  Using nested VFIO in a VM with less a few memory has show problems to get it working.
    **Status**: Missing more debug info
    Issue: kata-containers/tests#2677

Others:
    

- [ ]  #2829 virtiofsd crashes on Ubuntu 16.04 due to failing seccomp_init
- [ ] #2732 failed to launch qemu with virtio-fs configured bug needs-review
     **status**: low /dev/shm prevent launch kata
- [x]  #2555 random failures with virtiofs bug needs-review
- [x]  #2483 let virtiofsd and qemu run in parallel to speed up startup time of kata enhancement needs-review
- [x]  #2789 docker:dind not working in 1.11.1 qemu-virtiofs bug needs-review
- [x]  #2281 enable virtiofs when using kernel v4.19 bug needs-review
- [x]  #2137 kata-qemu-virtiofs fails to launch container bug needs-review

stress:
- [ ] #2813 virtiofsd crashes on kata container exit in k8s environment
       **status**: require reproducer
- [ ] #2888 run docker containers in a loop could have stability issues
      **status**: Not tested
- [ ]https://github.com/kata-containers/runtime/issues/2795

Metrics:
**Density**
- [ ] #2766 Less pods can be launched with virtio-fs as sharedfs bug needs-review
**Boot time**
- [ ] #2692 I am seeing very inconsistent launch times with kata-qemu-virtiofs from kata-deploy 1.11 bug needs-review
**I/O Performance**
- [ ] #2138 Poor qemu-virtiofs performance in benchmarks bug needs-review
- [ ] #2815 Virtiofs performance issue with huge files
**Memory**
- [ ]  KSM not helping because guest memory is marked as shared
    Status Seems that this is a limitation.
    kata-containers/runtime#2798
**CPU**
- [ ] Track and limit the amount of virtiofsd threads

enhancement:
- [ ] #2318 hypervisor: Create a virtiofsd process per each addDevice call enhancement needs-review
Maintainance:

**FAQ**:
Qemu virtiofsd is pointing to virtiofs fork, does qemu already has code merged to just use qemu in master?
```
Yes, virtiofs has been released as part of QEMU 5.0.
After upgrade to qemu 5.0 jobs has to point to that qemu binary
```
Where can I find that latest virtiofsd code ? From where kata should use virtiofsd code?
https://gitlab.com/virtio-fs/qemu/-/tags

Kata kernel for virtiofsd is pointing viritofsd fork with patches on top of it. 
Is it possible to just use upstream?
```
Yes, but there are missing patches to use DAX
```
What is the list missing patches not merged in linux upstream?
```
TODO: Fill list of patches (URLS)
```

**Overhead**:
   - What is the memory overhead of use virtiofsd, document CPU and memory using different virtiofs options, we need to find  and document best IO performance configuration, low memory overhead configuration, low CPU overhead configuration.
  - Make sure resize virtiofsd for small pods, for small pods we dont want that virtiofsd has a lot overhead (reduce amonut pool threads)

**Comment by fidencio** [comment]
@jcvenegas, virtiofs has been merged into QEMU 5.0.

**Comment by jcvenegas** [comment]
@fidencio thanks I updated the issue with that information

**Comment by c3d** [comment]
Summary of questions I asked about at today's kata-arch meeting:
* OverlayFS (use-case: making kata-containers work for container builds)
* inodes numbering (submount, sharing?)
* Shared directories / EmptyDir option 
* Resources usage, notably stack usage (@dagrh looking at it, comparison with 9p would be useful)
* `/dev` mounts (ran into a number of obscure code handling /dev last week)
* DAX support (my understanding is not ready yet)
* Cache size / caching defaults
* Attributes (passing `-o xattr` is fine as a workaround, can't make it a default I believe, can we enable it automatically in some cases, or will it be left to the user, if so documet)
* Related to the above: Privileged containers

**Comment by fidencio** [comment]
---

This issue is being automatically closed as Kata Containers 1.x has now reached EOL (End of Life). This means it is **no longer being maintained**.

> **Important:**
>
> All users should switch to the latest [Kata Containers 2.x release](https://github.com/kata-containers/kata-containers/blob/main/docs/install/README.md) to ensure they are using a maintained release that contains the latest security fixes, performance improvements and new features.

This decision was discussed by the @kata-containers/architecture-committee and has been announced via the [Kata Containers mailing list](http://lists.katacontainers.io/pipermail/kata-dev/):
* http://lists.katacontainers.io/pipermail/kata-dev/2020-November/001601.html
* http://lists.katacontainers.io/pipermail/kata-dev/2021-April/001843.html
* http://lists.katacontainers.io/pipermail/kata-dev/2021-May/001896.html

If you believe this issue still applies to Kata Containers 2.x, please open an issue against the [Kata Containers 2.x repository](https://github.com/kata-containers/kata-containers), pointing to this one, providing details to allow us to migrate it.

---
